wire-ios-data-model/Source/MLS/MLSDecryptionService.swift

```diff
@@ -121,42 +131,67 @@ public final class MLSDecryptionService: MLSDecryptionServiceInterface {
         }
 
         do {
-            let decryptedMessage = try await coreCrypto.perform {
-                try await $0.decryptMessage(
-                    conversationId: groupID.data,
-                    payload: messageData
-                )
-            }
+            let decryptedMessage = try await mlsActionExecutor.decryptMessage(messageData, in: groupID)
 
             if decryptedMessage.hasEpochChanged {
                 onEpochChangedSubject.send(groupID)
             }
 
-            if let commitDelay = decryptedMessage.commitDelay {
-                return MLSDecryptResult.proposal(commitDelay)
-            }
+            var results = try decryptedMessage.bufferedMessages?.compactMap({ try decryptResult(from: $0) }) ?? []
 
-            if let message = decryptedMessage.message {
-                return MLSDecryptResult.message(
-                    message,
-                    senderClientId(from: decryptedMessage)
-                )
+            if let result = try decryptResult(from: decryptedMessage) {
+                results.append(result)
             }
 
-            return nil
+            return results
         } catch {
             WireLogger.mls.error("failed to decrypt message for group (\(groupID.safeForLoggingDescription)) and subconversation type (\(String(describing: subconversationType))): \(String(describing: error))")
 
-            if case CryptoError.WrongEpoch(message: _) = error {
-                throw MLSMessageDecryptionError.wrongEpoch
-            } else {
+            switch error {
+            // Received messages targeting a future epoch, we might have lost messages.
+            case CryptoError.WrongEpoch: throw MLSMessageDecryptionError.wrongEpoch
+
+            // Message arrive in future epoch, it has been buffered and will be consumed later.
+            case CryptoError.BufferedFutureMessage: return []
+
+            // Received already sent or received message, can safely be ignored.
+            case CryptoError.DuplicateMessage: return []
+
+            // Received self commit, any unmerged group has know when merged by CoreCrypto.
```

Comment:
I'm a bit confused by this comment, is there a typo ?

Reply:
Yes

Reply:
Made it readable (I hope)